File: examples/cpp/flyweight.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

class CharacterStyle {
    std::string font_;
    int size_;
    std::string color_;
public:
    CharacterStyle(const std::string& font, int size, const std::string& color)
        : font_(font), size_(size), color_(color) {}

    void render(char c, int row, int col) const {
        std::cout << "  '" << c << "' at (" << row << "," << col
                  << ") [font=" << font_ << ", size=" << size_
                  << ", color=" << color_ << "]" << std::endl;
    }
};

class StyleFactory {
    std::unordered_map<std::string, std::shared_ptr<CharacterStyle>> styles_;
public:
    std::shared_ptr<CharacterStyle> getStyle(const std::string& font, int size, const std::string& color) {
        std::string key = font + "_" + std::to_string(size) + "_" + color;
        auto it = styles_.find(key);
        if (it == styles_.end()) {
            std::cout << "Creating new style: " << key << std::endl;
            auto style = std::make_shared<CharacterStyle>(font, size, color);
            styles_[key] = style;
            return style;
        }
        return it->second;
    }

    size_t styleCount() const { return styles_.size(); }
};

int main() {
    std::cout << "=== Flyweight: Character Style in Text Editor ===" << std::endl;

    StyleFactory factory;

    auto style1 = factory.getStyle("Arial", 12, "black");
    auto style2 = factory.getStyle("Arial", 12, "black");  // reused
    auto style3 = factory.getStyle("Times", 14, "red");

    std::cout << "\nRendering characters:" << std::endl;
    style1->render('H', 0, 0);
    style2->render('e', 0, 1);
    style3->render('l', 0, 2);

    std::cout << "\nTotal unique styles created: " << factory.styleCount() << std::endl;

    return 0;
}

File: examples/csharp/Patterns/Flyweight.cs
namespace DesignPatterns.Flyweight
{
    // Flyweight: shared character style
    public class CharacterStyle
    {
        public string Font { get; }
        public int Size { get; }
        public string Color { get; }

        public CharacterStyle(string font, int size, string color)
        {
            Font = font; Size = size; Color = color;
        }

        public override string ToString() => $"[{Font}, {Size}pt, {Color}]";
    }

    public class CharacterStyleFactory
    {
        private readonly Dictionary<string, CharacterStyle> _cache = new();

        public CharacterStyle GetStyle(string font, int size, string color)
        {
            var key = $"{font}_{size}_{color}";
            if (!_cache.ContainsKey(key))
            {
                _cache[key] = new CharacterStyle(font, size, color);
                Console.WriteLine($"  Created new style: {_cache[key]}");
            }
            return _cache[key];
        }

        public int StyleCount => _cache.Count;
    }

    public class Character
    {
        public char Char { get; }
        public CharacterStyle Style { get; }
        public int Row { get; }
        public int Col { get; }

        public Character(char c, CharacterStyle style, int row, int col)
        {
            Char = c; Style = style; Row = row; Col = col;
        }

        public override string ToString() =>
            $"'{Char}' at ({Row},{Col}) style={Style}";
    }

    public static class Demo
    {
        public static void Run()
        {
            var factory = new CharacterStyleFactory();
            var chars = new List<Character>();

            var text = "Hello";
            for (int i = 0; i < text.Length; i++)
                chars.Add(new Character(text[i], factory.GetStyle("Arial", 12, "black"), 0, i));

            var text2 = "World";
            for (int i = 0; i < text2.Length; i++)
                chars.Add(new Character(text2[i], factory.GetStyle("Arial", 12, "black"), 1, i));

            chars.Add(new Character('!', factory.GetStyle("Arial", 16, "red"), 1, 5));

            Console.WriteLine($"Total characters: {chars.Count}");
            Console.WriteLine($"Unique styles: {factory.StyleCount}");
        }
    }
}
